File: solutions/blind-75/linked-list/removeNthNodeFromEndOfList.js
const getNode = (head, n) => {
    let nthNode = head;
    
    for (let i = 0; i < n - 1; i++) {
        nthNode = nthNode.next;
    }
    
    return nthNode;
};

const deleteNextNode = (prevNode) => {
    prevNode.next = prevNode.next.next;
};

const removeNthFromEnd = (head, n) => {
    // We use a dummy node to avoid having to check if the node to remove is the head
    const sentinel = new ListNode(0, head);
    let nodeBeforeNthFromEnd = sentinel;

    let leadNode = getNode(head, n + 1);
    
     // We move the nth node to the end while also moving the other node the same number of times
    while (leadNode) {
        leadNode = leadNode.next;
        nodeBeforeNthFromEnd = nodeBeforeNthFromEnd.next;
    }
    
    deleteNextNode(nodeBeforeNthFromEnd);
    
    // We skip the dummy node to return the list
    return sentinel.next
};

File: solutions/python/blind-75/linked-list/remove_nth_node_from_end_of_list.py

class Solution:
    def get_node(self, head, n):
        nthNode = head
        
        for i in range(n - 1):
            nthNode = nthNode.next
        
        return nthNode
    
    def delete_next_node(self, prevNode):
        prevNode.next = prevNode.next.next
        
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # We use a dummy node to avoid having to check if the node to remove is the head
        sentinel = ListNode(0, head)
        
        # first and second pointer are n nodes apart
        firstPointer = sentinel
        secondPointer = self.get_node(head, n + 1)
        
        # We move both pointers until the second pointer reaches the end
        while secondPointer:
            firstPointer = firstPointer.next
            secondPointer = secondPointer.next
        
        # By this point, the first pointer will have n nodes in front of it
        self.delete_next_node(firstPointer)
        
        # We skip the dummy node to return the list
        return sentinel.next
    
    